### PR TITLE
identify: don't filter dns addresses based on remote addr type

### PR DESCRIPTION
**Pull request by sukunrt**
closes: #2550

### CODE REVIEW

**Review by marten-seemann — APPROVED** [review]
Do we need to make the same change in Kademlia?

Would it make sense to move all of this logic to go-multiaddr? It seems like every protocol that is accepting address advertisements from peers needs exactly the same filtering logic. (If so, we can do that in a follow-up PR, this code looks correct).

**Comment by marten-seemann** [comment]
> I think it makes sense to modify `multiaddr.IsPublicAddr` to return true for `/dns,/dns4/dns6` addresses, but only in cases where the name is not one of the special use dns names defined in [RFC 6761](https://datatracker.ietf.org/doc/html/rfc6761)

Makes sense.

File: p2p/protocol/identify/id_glass_test.go [hunk 1]
```diff
@@ -173,3 +174,34 @@ func TestInvalidSignedPeerRecord(t *testing.T) {
 	require.True(t, ok)
 	require.Nil(t, cab.GetPeerRecord(h2.ID()))
 }
+
+func TestIncomingAddrFilter(t *testing.T) {
+	lhAddr := ma.StringCast("/ip4/127.0.0.1/udp/123/quic-v1")
+	privAddr := ma.StringCast("/ip4/192.168.1.101/tcp/123")
+	pubAddr := ma.StringCast("/ip6/2::1/udp/123/quic-v1")
+	dnsAddr := ma.StringCast("/dns/example.com/udp/123/quic-v1")
```
Comment (marten-seemann) [inline p2p/protocol/identify/id_glass_test.go:182]: Should we add a `/dns/localhost` address here?

**Review by marten-seemann — APPROVED** [review]

File: p2p/protocol/identify/id_glass_test.go (hunk 1 above)
Comment (sukunrt) [reply p2p/protocol/identify/id_glass_test.go:182]: I'd prefer adding this when we start using the updated functions in [go-multiaddr](https://github.com/multiformats/go-multiaddr/pull/221) package.

Comment (marten-seemann) [reply p2p/protocol/identify/id_glass_test.go:182]: Here's the v0.12 release PR: https://github.com/multiformats/go-multiaddr/pull/223

Comment (marten-seemann) [reply p2p/protocol/identify/id_glass_test.go:182]: v0.12 was released: https://github.com/multiformats/go-multiaddr/releases/tag/v0.12.0

@sukunrt Can you update this PR?

**Comment by sukunrt** [comment]
sorry about the delay. I was procrastinating this thinking autorelay tests would fail, but turns out those tests don't actually use /dns4/localhost anywhere.

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**